### PR TITLE
ranking: Fix cascade behavior on `codeintel_ranking_exports`

### PR DESCRIPTION
**Pull request by efritz**
Ensure that `lsif_uploads` deletes do not cascade to `codeintel_ranking_exports` so that we can delete the associated files from GCS without losing the only reference.

## Test plan

Additional store unit tests.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (0.00 kb) | 0.00% (0.00 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/ef/fix-cascade-behavior/compare-report.html;tab=live_object) for a full comparison between the commits 83cb10b and e711dd7 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: internal/codeintel/codenav/internal/store/store.go [hunk 1]
```diff
@@ -163,7 +177,5 @@ FOR UPDATE OF re SKIP LOCKED
 
 const deleteStaleExportedUploadsQuery = `
 DELETE FROM codeintel_ranking_exports re
-WHERE
-	re.graph_key = %s AND
-	re.upload_id = ANY(%q)
+WHERE re.id = ANY(%q)
```
Comment (chrismwendt) [inline internal/codeintel/codenav/internal/store/store.go:180]: I haven't seen `%q` before, and I don't see any hits in the sqlf repo, so make sure that's what you intended.

CleanShot 2022-11-04 at 12 04 16@2x

File: migrations/frontend/1667497565_fix_cascade_behavior_for_codeintel_ranking_exports/up.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+-- Make column + foreign key nullable
+ALTER TABLE codeintel_ranking_exports DROP CONSTRAINT IF EXISTS codeintel_ranking_exports_upload_id_fkey;
+ALTER TABLE codeintel_ranking_exports ALTER COLUMN upload_id DROP NOT NULL;
+ALTER TABLE codeintel_ranking_exports ADD CONSTRAINT codeintel_ranking_exports_upload_id_fkey FOREIGN KEY (upload_id) REFERENCES lsif_uploads(id) ON DELETE SET NULL;
+
+-- Emulate old primary key index
+CREATE UNIQUE INDEX IF NOT EXISTS codeintel_ranking_exports_upload_id_graph_key ON codeintel_ranking_exports(upload_id, graph_key);
+
+-- TODO
```
Comment (efritz) [inline migrations/frontend/1667497565_fix_cascade_behavior_for_codeintel_ranking_exports/up.sql:14]:
```suggestion
Add prefix to control authoritative GCS paths as we might set upload_id to NULL now
```

File: migrations/frontend/1667497565_fix_cascade_behavior_for_codeintel_ranking_exports/down.sql [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+-- Restore the old primary key
+ALTER TABLE codeintel_ranking_exports DROP COLUMN IF EXISTS id;
+ALTER TABLE codeintel_ranking_exports ADD PRIMARY KEY (upload_id, graph_key);
+
+-- Drop now duplicate index
+DROP INDEX IF EXISTS codeintel_ranking_exports_upload_id_graph_key;
+
+-- TODO
+ALTER TABLE codeintel_ranking_exports DROP COLUMN IF EXISTS object_prefix;
```
Comment (chrismwendt) [inline migrations/frontend/1667497565_fix_cascade_behavior_for_codeintel_ranking_exports/down.sql:18]: TODO here

**Review by chrismwendt — APPROVED** [review]

File: migrations/frontend/1667497565_fix_cascade_behavior_for_codeintel_ranking_exports/down.sql (hunk 3 above)
Comment (efritz) [reply migrations/frontend/1667497565_fix_cascade_behavior_for_codeintel_ranking_exports/down.sql:18]: Updated :)

File: internal/codeintel/codenav/internal/store/store.go (hunk 1 above)
Comment (efritz) [reply internal/codeintel/codenav/internal/store/store.go:180]: IT IS NOT

**Pull request merged**